L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,504 @@
+L46: C-core: New TLS Credentials API
+----
+* Author(s): gtcooke94, ZhenLian
+* Approver: 
+* Status: Draft
+* Implemented in:
+* Last updated: May 16, 2024
+
+## Abstract
+
+This proposal aims to provide a new TLS credentials API with the following features:
+
+1. Credential reloading: reload transport credentials periodically from various sources, e.g. from credential files, or users' customized Provider implementations, without restarting the process.
+1. Customizable verification of the peer certificate: perform custom validation checks on the peer's certificate chain after it has been verified to chain up to a root of trust, or fully customize path building and cryptographic verification.
+1. TLS version configuration: optionally configure the minimum and maximum of the TLS versions to be negotiated during the handshake.
+1. Support for certificate revocation via Certificate Revocation Lists (CRLs).
+
+## Background
+
+The current TLS implementation (the SSL credentials API) in gRPC C core has some restrictions:
+1. Lack of the credential reloading support. Once an application starts, it can never reload its private key, certificate chain conveying the application's identity, and root certificates without shutting down.
+1. Inflexibility of how the peer certificate chain is verified. There is no API to add additional custom verification checks.
+1. Inability to choose the maximum/minimum TLS version to be negotiated during the TLS handshake.
+1. No support for certificate revocation.
+
+We propose an API that supports the above features and meets the following requirements:  
+
+1. Flexible enough to support multiple use cases. Users should be able to write their own reloading or verification logic if the provided ones do not fit their needs.
+1. The credential reloading should be efficient. We shall reload only when needed rather than reloading per TLS connection.
+1. C core API needs to be simple enough, so that most of existing SSL credentials in wrapped languages can migrate to call this new API.
```

Comment:
From the review: Possibly clean up "C core" reference.

Reply:
I think the piece of the puzzle that needs to be explained is that this is one of the first major features where the C-core layer is being written in C++, and this is the piece that needs emphasis. It looks like this is done below though, so resolving.